rpc: Expose g_is_mempool_loaded via getmempoolinfo

Empact opened the pull request:
And use it to fix a race condition in mempool_persist.py:
https://travis-ci.org/Empact/bitcoin/jobs/487577243

Since e.g. getrawmempool returns errors based on this status, this
enables users to test it for readiness.

Fixes #12863

Empact committed:
rpc: Expose g_is_mempool_loaded via getmempoolinfo and /rest/mempool/info.json

And use it to fix a race condition in mempool_persist.py:
https://travis-ci.org/Empact/bitcoin/jobs/487577243

Since e.g. getrawmempool returns errors based on this status, this
enables users to test it for readiness.

Empact commented:
The test failure, for reference:
```
stdout:
2019-02-01T20:25:18.578000Z TestFramework (INFO): Initializing test directory /tmp/test_runner_₿_🏃_20190201_201908/mempool_persist_74
2019-02-01T20:25:27.680000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/travis/build/Empact/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/util.py", line 113, in try_rpc
    fun(*args, **kwds)
  File "/home/travis/build/Empact/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/coverage.py", line 47, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/travis/build/Empact/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/authproxy.py", line 136, in __call__
    raise JSONRPCException(response['error'])
test_framework.authproxy.JSONRPCException: The mempool was not loaded yet (-1)
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/travis/build/Empact/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/test_framework.py", line 174, in main
    self.run_test()
  File "/home/travis/build/Empact/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/mempool_persist.py", line 130, in run_test
    assert_raises_rpc_error(-1, "Unable to dump mempool to disk", self.nodes[1].savemempool)
  File "/home/travis/build/Empact/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/util.py", line 105, in assert_raises_rpc_error
    assert try_rpc(code, message, fun, *args, **kwds), "No exception raised"
  File "/home/travis/build/Empact/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/util.py", line 119, in try_rpc
    raise AssertionError("Expected substring not found:" + e.error['message'])
AssertionError: Expected substring not found:The mempool was not loaded yet
2019-02-01T20:25:27.682000Z TestFramework (INFO): Stopping nodes
2019-02-01T20:25:27.985000Z TestFramework (WARNING): Not cleaning up dir /tmp/test_runner_₿_🏃_20190201_201908/mempool_persist_74
2019-02-01T20:25:27.985000Z TestFramework (ERROR): Test failed. Test logging available at /tmp/test_runner_₿_🏃_20190201_201908/mempool_persist_74/test_framework.log
2019-02-01T20:25:27.985000Z TestFramework (ERROR): Hint: Call /home/travis/build/Empact/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/combine_logs.py '/tmp/test_runner_₿_🏃_20190201_201908/mempool_persist_74' to consolidate all logs
stderr:
```

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -1508,6 +1509,7 @@ static UniValue getmempoolinfo(const JSONRPCRequest& request)
                 {},
                 RPCResult{
             "{\n"
+            "  \"loaded\": true|false         (boolean) True if the mempool has been loaded/initialized.\n"
```
Should clarify that this means loaded *from disk* and if -persistmempool is on?

Empact replied:
If `-persistmempool` is off, `g_is_mempool_loaded` is set at the end of `ThreadImport`, after block import. Ideas on how to articulate that?

Empact replied:
How about this: https://github.com/bitcoin/bitcoin/pull/15323/commits/fce31d4d96704dde653e2c19091ff2e703f28323

promag replied:
nit, why `/initialized`?

Empact replied:
Because it will also be true in cases where it has not been loaded from disk, but only "initialized". Maybe not worth mentioning.

laanwj replied:
Would prefer something like
`True if the mempool has been loaded from disk (if persistent)`

jnewbery replied:
I think `True if the mempool has been loaded from disk (if persistent), or need not be` is much less clear than the more simple `True if the mempool is fully loaded.` The first raises more questions than it answers ("What is persistent? When need it not be?")

MarcoFalke replied:
Maybe `s/persist/-persistmempool/g`?

MarcoFalke commented:
Ref https://github.com/bitcoin/bitcoin/issues/12863#issuecomment-381031762

promag reviewed (commented):
Concept ACK. Could add a release note with the new key. This introduces a weird case, which is `loaded` can change from `true` to `false`. Maybe
https://github.com/bitcoin/bitcoin/blob/b3a715301a0fd972fb2f3bd36e2680b3cdbbab26/src/init.cpp#L707
should change to
```cpp
 g_is_mempool_loaded |= !ShutdownRequested(); 
```

promag commented:
`/rest/mempool/info.json` is also affected.

Empact commented:
Revised and cleaned up:
* Updated `/rest/mempool/info.json` to include the attribute (thanks @promag)
* Split the setting of `g_is_mempool_loaded` strictly between the `persistmempool` and non-persist case - note it's only relevant on shutdown if `persistmempool` is set: https://github.com/bitcoin/bitcoin/blob/8f470ecc534d49cb5f7045ea9564481acd3e2b4b/src/init.cpp#L238-L240

Empact commented:
Updated comments and documentation as per John and Marco's suggestions https://github.com/bitcoin/bitcoin/pull/15323#discussion_r260384720

Empact committed:
Move g_is_mempool_loaded into CTxMemPool::m_is_loaded

So the loaded state is explicitly mempool-specific.

Empact commented:
Rebase for #15473, and add another commit moving `g_is_mempool_loaded` into `CTxMemPool`, given that it is mempool-specific.

jnewbery commented:
ACK the first and third commits (_rpc: Expose g_is_mempool_loaded via getmempoolinfo and /rest/mempool/info.json_ and _Move g_is_mempool_loaded into CTxMemPool::m_is_loaded_). I like the tidy up in _Move g_is_mempool_loaded into CTxMemPool::m_is_loaded_ to remove the `g_is_mempool_loaded` global.

I don't see the need for the second commit (_rpc: Set g_is_mempool_loaded early if the mempool is not persisted_). Having two places where `m_is_loaded_` gets set seems to add complexity for no good reason. I don't think any users will see any behaviour difference before and after that commit.

MarcoFalke commented:
utACK effe81f750

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

utACK effe81f750
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgQDgv/Vu+DeQYZkt0Au46dAPXOWF5uIfLHOGhOHbKB+YZ6zFNIrmw6VF5XrYDl
IJqLlqX2WtDQd8gYcQ9dDvVSBktaJrk6Qm9laEX8qITMavz5VHK8rHfqmJ9nrqC8
DNgWMlwhoQsJqflqXeaUKnkXv9eoMLvuHVn+F3kDkr9Wu/AjRjLxu1zaE9qnr0sl
lO8Z+mdVig4NDWbCHwZHyvP+eM8skn5WeEXTSi6Vl1yqgEP9T/9TEFQtlQ1ikTjn
t+ACMSxzfest155gsZt0vxn5ABlfw1phw3b/x6DEFIFNzs0cBYVRxu8kPhTarqYL
gMcrAvVlfs9knGDUaU4XaZbfor/cQRAduFCKqowSxob4jwb2FcWX9jDCxuyCBge7
kPS1qFNI0eBRw0OLSKKq3A5wH8Oyvx5XzagoRkFcKCz2hF1KqFfLKT+EyBViBWbR
lW6qMTO+A26VJsslpzaTpfnQrTU3mfyUxJB3+IPVQ0Mcibt2jBdfYJCq0/W+DIGe
RtTjJIFt
=u6Rr
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `fcca790e597327e6dbf958e78e5e5c02c0911ef653160c591bf56423b214127d  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108fcca790e597327e6dbf958e78e5e5c02c0911ef653160c591bf56423b214127df010443413b8fa53226e99a7dd899f59ab0208fff010784fda0e08d2d26a703306d8c54babbc08f1045cc8cfaaf008d9d1ec1de408b5270083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010a27224b3d4666824fdd8651074d2020d08f1045cc8cfa9f008a7bfe762643e3a7a0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010af49883cc0f1c781537280de470f8b3208f1045cc8cfa9f0082ddfaef2725a66750083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f4bc54af7225ec88324deccf95a6e3ef08f010c8969c6fb41239c8bb57a293036631a508f1045cc8cfaaf008ac1e911b3473d5d30083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

jnewbery commented:
utACK effe81f7503d2ca3c88cfdea687f9f997f353e0d

@Empact - apologies for not reACKing this earlier. Force-pushes don't trigger new notifications in github. Feel free to ping me or add a comment to the PR if you need a reACK
